### DISCUSSION TITLE
edge functions which can be invoked only by the project creator or by internally(i.e. db webhooks)

### DISCUSSION THREAD

**Discussion by serhatdirihan** [supabase/supabase#22113] [Questions]
Hello,

Do you know if there is a setting on the edge functions which allows it to be invoked only by the project creator, or only by internally via db webhooks? 

The reason why I need such requirement is that, I follow the tutorial regarding calling push notifications via edge functions, it sends the push notification by db webhook when there is an insert to a table. It works fine. However, if someone else knows the name of the edge function, he can easily invoke it and able to trigger push notifications(especially in that example the edge function was using SUPABASE_SERVICE_ROLE_KEY). 

Another reason, imagine I created an edge function and I want to invoke it from an external app based on some calculations, eventually the edge function is going to insert a record to a table. I would like only myself(as the project creator) to call this edge function to insert the record, and all the users of the project can only read that table. If I use SUPABASE_SERVICE_ROLE_KEY in the edge function, then if someone somehow knows the name of the edge function, he can easily trigger it. As far as I know either I have to create a user for me in the project(sort of admin user), and fetch the user id from the token in the edge function and check if it is my user or not before inserting the record, or I need to use third party solutions like https://unkey.dev/ and send an API KEY when I call the edge function and verify it in the edge function before inserting the record. But, I would like to see if supabase has some features to achieve this out of the box. 

Any suggestions?

Thank you

**Comment by bahaafathi** [comment]
@serhatdirihan  i have the same problem, did you find any solution

**Comment by pourya7** [comment]
Facing the same issue here!

I'm following the [Semantic Search tutorial](https://supabase.com/docs/guides/functions/examples/semantic-search) to test out the AI features, where I need to have a server function `generate-embedding` invoked by a webhook each time there is new content added or updated.

Then I followed the [Database Webhooks tutorial](https://supabase.com/docs/guides/database/webhooks) to setup my webhook that calls the edge function to generate the embeddings. But since the edge function uses the `SUPABASE_SERVICE_ROLE_KEY`, if someone figures out the function name, they can invoke it.

Trying to find a way to limit access to this function, I have added an extra header `X-SECRECT-KEY` to the webhook, that can later be validated on the edge function, but having `SUPABASE_ANON_KEY` and the `X-SECRECT-KEY` makes the code less secure. We don't want to have these keys as a part of our migration code.

**Reply by pourya7** [reply]
The answer from this post helped me fix the issue:
https://stackoverflow.com/a/78900192/4281702

By creating a edge function wrapper for your webhooks (alternative to `supabase.http_request`), you can add secret headers to your internal function calls. The secrets can easily be stored on Vault (on both local and prod).

Later, in the edge runtime you can retrieve env variables and compare them with the headers coming from the internal http requests, and reject unauthorised requests.

To set edge runtime env variables in local environment, run:
`supabase functions serve --env-file .\.env.local`

To set edge runtime env variables in self-hosted environment, just add them under `docker-compose.yml` file under `functions` sections.
